pubsub/pubsub.go

```diff
@@ -211,12 +213,38 @@ type Subscription struct {
 	ackBatcher driver.Batcher
 	cancel     func() // for canceling all SendAcks calls
 
-	mu    sync.Mutex    // protects everything below
-	q     []*Message    // local queue of messages downloaded from server
-	err   error         // permanent error
-	waitc chan struct{} // for goroutines waiting on ReceiveBatch
+	mu                sync.Mutex    // protects everything below
+	q                 []*Message    // local queue of messages downloaded from server
+	err               error         // permanent error
+	waitc             chan struct{} // for goroutines waiting on ReceiveBatch
+	lastReceiveReturn time.Time     // time that the last call to Receive returned
+	avgProcessTime    float64       // moving average of the seconds to process a message
 }
 
+const (
+	// The desired length of a subscription's queue of messages (the messages pulled
+	// and waiting in memory to be doled out to Receive callers). The length is
+	// expressed in time; the relationship to number of messages is
+	//
+	//      lengthInMessages = desiredQueueLength / averageProcessTimePerMessage
+	//
+	// In other words, if it takes 100ms to process a message on average, and we want
+	// 2s worth of queued messages, then we need 2/.1 = 20 messages.
+	//
+	// If desiredQueueLength is too small, then there won't be a large enough buffer
+	// of messages to handle fluctuations in processing time, and the queue is likely
+	// to become empty, reducing throughput. If desiredQueueLength is too large, then
+	// messages will wait in memory for a long time, possibly timing out (that is,
+	// their ack deadline will be exceeded). Those messages could have been handled
+	// by another process receiving from the same subscription.
+	desiredQueueLength = 2 * time.Second
```

Comment:
Are you okay with desiredQueueDuration?

Reply:
Sure, as long as the comment explains what it means. I think it's how long we want a message to spend in the queue, although I find this idea a bit strange since I don't have a preference about how long messages stay in the queue so long as it's a lot less than the ack deadline.

Reply:
> although I find this idea a bit strange

Exactly, that's why I don't think that's the right way to think about it. It's more like, we want to keep a few messages around in case Receive speeds up and starts chewing through them faster than it has been. In other words, a buffer. How many messages do we want to keep in the buffer? No, that's the wrong question: how much runway (in time) do we want before we run out of messages?

Yes, if Receive always takes the same time, then a message will spend all that time in the queue. But that's a side effect, the price we pay for having messages available on demand to improve throughput.

Reply:
I see, so to write it in Javanese, it would be more like `desiredAmountOfTimeBeforeQueueProbablyRunsOut`. Makes sense.